tests/TestCase/ORM/QueryTest.php

```diff
@@ -3688,4 +3688,39 @@ public function testSelectAllExceptThrowsInvalidArgument()
                 ->find()
                 ->selectAllExcept([], ['body']);
     }
+
+    /**
+     * Test that the entity of a belongsTo association is not empty if we are using an inner join with match
+     * and specify a list of select fields
+     */
+    public function testInnerJoinMatchWithQueryFields()
+    {
+        $table = $this->getTableLocator()->get('Articles');
+        $table->belongsTo('Authors');
+
+        // 1. without providing a set of select fields the test succeeds
+        $query = $table->find('all', [
+            'contain' => ['Authors']
+        ]);
+        $query->innerJoinWith('Authors', function ($q) {
+            return $q->where(['Authors.name' => 'larry']);
+        });
+        $article = $query->first();
+        $this->assertNotEmpty($article->author);
+
+        // 2. when providing an array of fields to select the author-entity is empty
+        $query = $table->find('all', [
+            'contain' => ['Authors']
+        ]);
+        $query->innerJoinWith('Authors', function ($q) {
+            return $q->where(['Authors.name' => 'larry']);
+        });
```

Comment:
You don't need to mix `contain` + `innerJoinWith`. You can use `contain()` and `where()`

Reply:
You're right, and I fixed the problem in the project I got this from by replacing the `innerJoinWith` with `contain` and `where`.
Do you think that it might nonetheless be worth it to look into why the author-entity goes missing, when using this particular combination of `contain`, `innerJoinWith` and `select`? I don't want to waste my or anybodies time with it.

Reply:
Figuring out why is a complicated time-intensive process that I don't have the energy for.

Reply:
Before you figure out why and a potential fix, it might be good to determine what the correct behavior should be. Right now I think matching will overrule contain. Scenarios that might make the decision less simple are how it behaves when aggregate fields are in use.